### PR TITLE
[Proposal] Run CS Tests in a specific 5.6 build + disable xdebug for cs tests (and reorder execution matrix)

### PR DESCRIPTION
**Pull request by andrepereiradasilva**
### Summary of Changes

Run CS Tests in a specific 5.6 build to improved speed of travis checking.
Also reorders execution matrix for faster tests (nodejs to last, cs to first).
Also disables xdebug is cs tests for faster performance.
### Improvements
- **Mandatory tests time**: From 8:00 min to 4:45 min (aroud 40% improvement)
- **All tests time**: From 20:00 min to 13:00 min (aroud 35% improvement)
- **CS Results in**: 7:00/8:00 min to 2:00/3:00 min (more than 60% improvement)
#### Before

image
#### After

image
### Testing Instructions

Unit test + cs tests pass.
Code review.
### Documentation Changes Required

None.

### CODE REVIEW

**Comment by andrepereiradasilva** [comment]
@photodude @mbabker what do you think?

**Comment by andrepereiradasilva** [comment]
BTW this is also as the advantage that the CS checks are the first to be run so in no more than 4 minutes (instead of aprox 8 minutes before) we have the CS test results. Where actually most times travis fails.

**Comment by zero-24** [comment]
Looks good and travis is happy too.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11757">issues.joomla.org/joomla-cms/11757</a>.</sub>

**Comment by andrepereiradasilva** [comment]
@zero-24 now i disabled xdebug in cs tests let's see how much will improve further.

**Comment by andrepereiradasilva** [comment]
nice even faster 3 min and you have the cs results :)
image

**Comment by photodude** [comment]
Probably a good choice to run the code standards checks separate from the unit tests.

**Comment by mbabker** [comment]
We can probably disable XDebug on PHPUnit too because we aren't generating
code coverage on Travis.  And it may help with PHP 7.1's build segfaulting.

**Comment by andrepereiradasilva** [comment]
ok php debug disabled in unit tests too. testing ...

File: build/travis/unit-tests.sh [hunk 1]
```diff
@@ -10,6 +10,9 @@ set -e
 # Make sure all dev dependencies are installed
 composer install
 
+# Disable xdebug on php 7.0.* and lower.
+if [[ ( $TRAVIS_PHP_VERSION = 5.* ) || ( $TRAVIS_PHP_VERSION = 7.0.* ) ]]; then phpenv config-rm xdebug.ini; fi
```
Comment (photodude) [inline build/travis/unit-tests.sh]: for [hhvm you should be able to disable xdebug as follows](https://docs.hhvm.com/hhvm/configuration/INI-settings)
`if [[ $TRAVIS_PHP_VERSION = hhv* ]]; then echo  xdebug.enable=0  >> /etc/hhvm/php.ini; fi`

**Comment by photodude** [comment]
Since you are disabling xdebug for everything now IMO the `cs-tests.sh` file is no longer needed.

**Comment by andrepereiradasilva** [comment]
now time is like this:
image

So in aroud 5 minute swe have all mandotory tests done :)

**Comment by photodude** [comment]
On a side note: the total test time is a skewed value and is highly dependent on the hhvm test and is highly variable. Tests will be marked complete long before the total test time completes since the matrix build is marked complete when the build jobs pass even though the allowed failure tests are still running. 

Our actual test start to completion is just the longest time listed in the build jobs portion of the Travis CI matrix.

**Comment by yvesh** [comment]
Good work @andrepereiradasilva

File: build/travis/unit-tests.sh (hunk 1 above)
Comment (photodude) [inline build/travis/unit-tests.sh]: @andrepereiradasilva would you add the fix for disabling xdebug in hhvm?

**Comment by andrepereiradasilva** [comment]
> On a side note: the total test time is a skewed value and is highly dependent on the hhvm test and is highly variable. Tests will be marked complete long before the total test time completes since the matrix build is marked complete when the build jobs pass even though the allowed failure tests are still running.
> Our actual test start to completion is just the longest time listed in the build jobs portion of the Travis CI matrix.

I know that, but as long as the hhvm test don't finish it's not released to do other tests in other PR in the queue ...

**Comment by photodude** [comment]
It should be doing all the tests in the build matrix and marking the build complete before testing the allowed failures; unless travis isn't fully following their build matrix and documentation about the "fast finish" directive.

**Comment by andrepereiradasilva** [comment]
i don't meand the tests in the same commit i mean when you do another commit that is added to the tests queue.

**Comment by rdeutz** [comment]
I tried to test this but the xdebug disable doesn't work, anything I missed? Here is the travis-job: https://travis-ci.org/rdeutz/joomla-cms/jobs/154519045

**Comment by andrepereiradasilva** [comment]
See https://travis-ci.org/rdeutz/joomla-cms/jobs/154519045#L265

**Comment by rdeutz** [comment]
ahh but why not removing earlier not needed for composer

**Comment by rdeutz** [comment]
played around with it a bit seems you can't disable xdebug for the composer self-update but for anything else. I also tried to let all 5.6 together, because from I have seen is that travis starts 5 jobs and then the next one when one is finished. With two 5.6 jobs it took 17 minutes and in combination 13 miniutes. Can't really say what is better because I don't have so much tests run, so we might need to check both configurations.

@andrepereiradasilva that's is a good improvement, great job!

**Comment by yvesh** [comment]
@andrepereiradasilva imo you can remove xdebug always no reason for that weird if ;-)

**Comment by photodude** [comment]
@andrepereiradasilva would you disable Xdebug in [the hhvm unit tests as follows](https://docs.hhvm.com/hhvm/configuration/INI-settings)
`if [[ $TRAVIS_PHP_VERSION = hhv* ]]; then echo  xdebug.enable=0  >> /etc/hhvm/php.ini; fi`

**Comment by rdeutz** [comment]
I have created a remove xdebug script and called it Branch: https://github.com/rdeutz/joomla-cms/tree/test11757

**Comment by photodude** [comment]
@yvesh the if is necessary since HHVM and PHP 7.1 will error as they have a different format to configure the environment settings.

**Comment by andrepereiradasilva** [comment]
ok will add it

**Comment by andrepereiradasilva** [comment]
@mbabker it seems php 7.1 doesn't have xdebug enabled.

did a test to check the php conf in 7.1
```bash
ls -la ~/.phpenv/versions/$(phpenv version-name)/etc/
cat ~/.phpenv/versions/$(phpenv version-name)/etc/php.ini
ls -la ~/.phpenv/versions/$(phpenv version-name)/etc/conf.d/
cat ~/.phpenv/versions/$(phpenv version-name)/etc/conf.d/travis.ini
```
And didn't found any xdebug.

**Comment by andrepereiradasilva** [comment]
Latest test result: We are now below 5min to do all required tests :smile: 

image

**Comment by mbabker** [comment]
Well, that's interesting, since it's there by default on the other PHP builds IIRC.  Still doesn't explain now why PHPUnit segfaults on the Travis environment, but I guess that's something to bang my head over on the existing issue.

**Comment by andrepereiradasilva** [comment]
ok. so @wilsonge @rdeutz can this be merged so all tests run faster?
We can do more improvements later ...

**Comment by photodude** [comment]
I'm still of the option that the added `cs.tests.sh` File is unnecessary since that change was added to the original  bash build file.

**Comment by andrepereiradasilva** [comment]
ok let's see if it works like this (file cs-tests.sh removed)

**Comment by andrepereiradasilva** [comment]
ok so i reverted back to have the cs plus unit in 5.6 in one go. i think it's better now

**Comment by andrepereiradasilva** [comment]
hum ... it seems it's faster to have one php 5.6 for the unit tests.

If tests of the latest commit pass IMO is ready to merge.

**Comment by andrepereiradasilva** [comment]
thabjs @rdeutz finnaly we don't have to wait almost 10 minutes for the cs results!

**Pull request merged**